### PR TITLE
DAOS-5605 common: add basic infrastructure for QAT

### PR DESCRIPTION
**Pull request by johannlombardi**
Introduce interface for compression and cipher algorithms as well as
container properties to support compression and encryption.
Introduce a new interface just for hash functions (called multihash
since hash is already used by Gurt for hash table) and make it
independent of the csummer interface. All ISA-L functions are moved
to a specific file so that we can eventually support different
implementations (e.g. QAT).
Remove mchecksum support since it is not used any longer.

Signed-off-by: Johann Lombardi <johann.lombardi@intel.com>

### CODE REVIEW

File: src/common/multihash_isal.c [hunk 1]
```diff
@@ -0,0 +1,444 @@
...
+	*crc16 = crc16_t10dif(*crc16, buf, (int)buf_len);
+	return 0;
+}
+
+static int
+crc16_finish(void *daos_mhash_ctx, uint8_t *buf, size_t buf_len)
+{
+	uint16_t *crc16 = (uint16_t *)daos_mhash_ctx;
+
+	*((uint16_t *)buf) = *crc16;
```
Comment (johannlombardi) [inline src/common/multihash_isal.c:81]: @ryon-jensen Could you please advise whether we need to auto-reset on finish?

File: src/common/multihash_isal.c [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+ * ---------------------------------------------------------------------------
+ * ISA-L Hash Algorithms
+ * ---------------------------------------------------------------------------
+ */
+
+/** CRC16_T10DIF*/
+static int
+crc16_init(void **daos_mhash_ctx)
+{
+	D_ALLOC(*daos_mhash_ctx, sizeof(uint16_t));
```
Comment (johannlombardi) [inline src/common/multihash_isal.c:47]: @ryon-jensen I know that this is a bummer to allocate memory for this. The issue is that we have no proper storage via the multihash interface. The other option would be to have an interface for the user to pass a buffer and store it in the daos_mhash_ctx pointer. Would you have preferred this option instead? Any idea suggestions?

File: src/common/prop.c [hunk 3]
```diff
@@ -342,13 +342,19 @@ daos_prop_valid(daos_prop_t *prop, bool pool, bool input)
 			break;
 		case DAOS_PROP_CO_CSUM_CHUNK_SIZE:
 			/** Accepting anything right now */
+			val = prop->dpp_entries[i].dpe_val;
+			if (val >= (1ULL << 32)) {
```
Comment (ryon-jensen) [inline src/common/prop.c:346]: I would just remove comment above as error message provides enough info. But isn't 1<<32 4GiB?

File: src/common/multihash_isal.c [hunk 4]
```diff
@@ -0,0 +1,444 @@
+/**
+ * (C) Copyright 2019-2020 Intel Corporation.
```
Comment (ryon-jensen) [inline src/common/multihash_isal.c:2]: ditto

**Review by ryon-jensen — COMMENTED** [review]
Overall it looks good. Just a few small comments.

- I'm assuming more will go into the cipher and compression interfaces.
- Should we use a different facility (D_LOGFAC) for cipher.c, multihash.c, and compression.c? They're all related but not sure csum is right.

File: src/common/multihash_isal.c (hunk 2 above)
Comment (ryon-jensen) [reply src/common/multihash_isal.c:47]: I think this way is acceptable. I tried to reuse the same csummer as much as possible for a container on both the server and client so init is called as infrequently as possible.

File: src/common/multihash_isal.c (hunk 1 above)
Comment (ryon-jensen) [reply src/common/multihash_isal.c:81]: For now it isn't expected that finish auto-resets, though maybe something we improve in the future.

File: src/common/checksum.c [hunk 5]
```diff
@@ -38,6 +36,9 @@
 #define C_TRACE(...) D_DEBUG(DB_CSUM, __VA_ARGS__)
 #define C_TRACE_ENABLED() D_LOG_ENABLED(DB_TRACE)
 
+/** ISA-L hash function table implemented in multihash_isal.c */
+extern struct hash_ft *isal_hash_algo_table[];
```
[the remark this answers is not in the document]
Comment (johannlombardi) [reply src/common/checksum.c:40]: nope, good catch. Just leftover from code move.

File: src/common/prop.c (hunk 3 above)
Comment (johannlombardi) [reply src/common/prop.c:346]: Yes, it is encoded on 32 bits in the pool service. Any concerns? We can impose a lower value if you prefer.

Comment (ryon-jensen) [reply src/common/prop.c:346]: No, I think that's fine. The error message below says "Should be < 2GiB" which confused me.

**Review by ryon-jensen — APPROVED** [review]

**Review by bfaccini — APPROVED** [review]

**Pull request merged**